### PR TITLE
add sts error metric

### PR DESCRIPTION
**Pull request by jyotimahapatra**
Signed-off-by: Jyoti Mahapatra <jyotima@amazon.com>

### CODE REVIEW

File: pkg/token/token.go [hunk 1]
```diff
@@ -514,6 +515,7 @@ func (v tokenVerifier) Verify(token string) (*Identity, error) {
 		if urlErr, ok := err.(*url.Error); ok {
 			return nil, NewSTSError(fmt.Sprintf("error during GET: %v", urlErr.Err))
```
Comment (wongma7) [inline pkg/token/token.go:516]: why not inc here?

Comment (jyotimahapatra) [reply pkg/token/token.go:516]: right..should do here too...sending an update

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -41,6 +43,20 @@ func CreateMetrics(reg prometheus.Registerer) Metrics {
...
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "sts_connection_failures",
+				Help:      "Sts call could not succeed or timedout",
+			},
+		),
+		StsResponseError: factory.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "sts_failures",
```
Comment (wongma7) [inline pkg/metrics/metrics.go:56]: wat about name sts_request_failures instead? something more specific

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -41,6 +43,20 @@ func CreateMetrics(reg prometheus.Registerer) Metrics {
...
 			},
 		),
+		StsConnectionFailure: factory.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "sts_connection_failures",
+				Help:      "Sts call could not succeed or timedout",
+			},
+		),
+		StsResponseErrorCodeNot200: factory.NewCounter(
```
Comment (nckturner) [inline pkg/metrics/metrics.go:53]: I think we should have a metric called `sts_responses_total` and use labels for the exact response code.  This is in line with the recommendations here https://prometheus.io/docs/practices/instrumentation/#use-labels. This will also be much more useful in the long run then only knowing if the code is not 200.

File: pkg/metrics/metrics.go [hunk 4]
```diff
@@ -41,6 +43,20 @@ func CreateMetrics(reg prometheus.Registerer) Metrics {
 				Help:      "EKS Configmap watch failures",
 			},
 		),
+		StsConnectionFailure: factory.NewCounter(
+			prometheus.CounterOpts{
+				Namespace: Namespace,
+				Name:      "sts_connection_failures",
```
Comment (nckturner) [inline pkg/metrics/metrics.go:49]:
```suggestion
				Name:      "sts_connection_failures_total",
```

File: pkg/token/token_test.go [hunk 5]
```diff
@@ -152,7 +160,7 @@ func TestSTSEndpoints(t *testing.T) {
 }
 
 func TestVerifyTokenPreSTSValidations(t *testing.T) {
-	b := make([]byte, maxTokenLenBytes+1, maxTokenLenBytes+1)
```
Comment (nckturner) [inline pkg/token/token_test.go:155]: Just curious, what's the reason for removing capacity?

File: pkg/token/token.go [hunk 6]
```diff
@@ -564,7 +567,7 @@ func (v tokenVerifier) Verify(token string) (*Identity, error) {
 func hasSignedClusterIDHeader(paramsLower *url.Values) bool {
 	signedHeaders := strings.Split(paramsLower.Get("x-amz-signedheaders"), ";")
 	for _, hdr := range signedHeaders {
-		if strings.ToLower(hdr) == strings.ToLower(clusterIDHeader) {
```
Comment (nckturner) [inline pkg/token/token.go:567]: Nit: This seems like an improvement but should be in a different PR

File: pkg/token/token_test.go (hunk 5 above)
Comment (jyotimahapatra) [reply pkg/token/token_test.go:155]: the IDE told me to? it was redundant

Comment (jyotimahapatra) [reply pkg/token/token_test.go:155]: i can get rid of it for now

File: pkg/token/token.go (hunk 6 above)
Comment (jyotimahapatra) [reply pkg/token/token.go:567]: i can get rid of it for now

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (jyotimahapatra) [reply pkg/metrics/metrics.go]: I was following the metric convention from https://www.envoyproxy.io/docs/envoy/latest/configuration/upstream/cluster_manager/cluster_stats due to my familiarity. We can use labels here

File: pkg/token/token.go [hunk 7]
```diff
@@ -523,6 +525,7 @@ func (v tokenVerifier) Verify(token string) (*Identity, error) {
 		return nil, NewSTSError(fmt.Sprintf("error reading HTTP result: %v", err))
 	}
 
+	metrics.Get().StsResponseErrorCode.WithLabelValues(fmt.Sprint(response.StatusCode)).Inc()
```
Comment (wongma7) [inline pkg/token/token.go:528]: but do we care about 200s?  is 200 an error code? why is the metric name sts_responses_total if the variable name is sts_response_error_code (it's hard to tell intention from the naming)

Comment (nckturner) [reply pkg/token/token.go:528]: Yeah we should probably either rename `StsResponseErrorCode` or only increment for `response.StatusCode != 200`.  The benefit of including successes is the ability of getting success % metric.

Comment (nckturner) [reply pkg/token/token.go:528]: So the first option would be to just rename `StsResponseErrorCode` to `StsResponseCode`.

Comment (jyotimahapatra) [reply pkg/token/token.go:528]: yes..i missed that

Comment (jyotimahapatra) [reply pkg/token/token.go]: done..changed the variable name

**Comment by nckturner** [comment]
@jyoti please fix the merge conflict

**Comment by jyotimahapatra** [comment]
@nckturner ..rebased and ready
